### PR TITLE
allocate snopt memory to the heap (instead of the stack).

### LINKED ISSUE #933: need avoid large static allocations (e.g. in the c++ inverseKinematics / snopt interface)
just helped someone debug a mysterious failure running ik on windows, which was reporting as a STACK/HEAP errors with calls in Eigen::internal::malloc_aligned on the stack.  Turns out that it was because we were exceeding the default stack limit used by msvc on windows (which is 1MB).  

I followed the instructions here: https://social.msdn.microsoft.com/Forums/en-US/2a5b32b6-683b-4729-92d3-45ed7a89ef3f/stack-overflow-and-how-to-increase-the-stack-size?forum=Vsexpressvc

and set my stack size to 8000000.  then everything ran just fine.

recording this here in case it comes up again, but also because I think we should be more careful to avoid allocating big memory to the stack (with static allocation) instead of the heap (with new/malloc) in our code.

### PR DESCRIPTION
**Pull request by RussTedrake**
resolves #933 .  for more information, consider checking out 
http://gribblelab.org/CBootcamp/7_Memory_Stack_vs_Heap.html

### CODE REVIEW

**Comment by RussTedrake** [comment]
tried looking at the cdash history of timing for drake006 and drake005 for some of the snopt-intensive test.  the timings are so stochastic that it's really hard to figure out... it looks like a few of the trajectory optimization tests might be slightly faster now, and the IK tests might be slightly slower.  will do some quick timing analysis on this machine to see if I can get more repeatable numbers.

**Comment by RussTedrake** [comment]
Here are some timing results for a test that's essentially one big call to snopt (one of the littledog gait optimization examples).  The actual allocation at the beginning should be a small part of the runtime, so this is meant to evaluate whether accessing the memory has any effect (on my mac laptop, running on battery power...)

```
>> for i=1:10, gaitOptimization; end % with heap alloc
Elapsed time is 59.079900 seconds.
Elapsed time is 58.810267 seconds.
Elapsed time is 58.825587 seconds.
Elapsed time is 58.890060 seconds.
Elapsed time is 58.824613 seconds.
Elapsed time is 58.783372 seconds.
Elapsed time is 59.092312 seconds.
Elapsed time is 58.781579 seconds.
Elapsed time is 58.890583 seconds.
Elapsed time is 58.932540 seconds.
>> megaclear
>> for i=1:10, gaitOptimization; end % master
LCM: Disabling IPV6 support
LCM: TTL set to zero, traffic will not leave localhost.
Elapsed time is 61.019614 seconds.
Elapsed time is 60.155278 seconds.
Elapsed time is 59.821419 seconds.
Elapsed time is 60.049155 seconds.
Elapsed time is 59.866713 seconds.
Elapsed time is 60.040203 seconds.
Elapsed time is 60.155956 seconds.
Elapsed time is 63.661460 seconds.
Elapsed time is 63.434205 seconds.
Elapsed time is 59.911375 seconds.
```
and the other relevant test is something that goes into and out of snopt a LOT of times, like IK.  so i picked (at random) one of the atlas tests in the middle of systems/plant/testIK and put a for loop around it
```
Check IK pointwise with gaze target constraint  % with heap alloc
Elapsed time is 0.373562 seconds.
Elapsed time is 0.352501 seconds.
Elapsed time is 0.375544 seconds.
Elapsed time is 0.377052 seconds.
Elapsed time is 0.353330 seconds.
Elapsed time is 0.355910 seconds.
Elapsed time is 0.367895 seconds.
Elapsed time is 0.351513 seconds.
Elapsed time is 0.352541 seconds.
Elapsed time is 0.349272 seconds.
Elapsed time is 0.370571 seconds.
Elapsed time is 0.359265 seconds.
Elapsed time is 0.359097 seconds.
Elapsed time is 0.361806 seconds.
Elapsed time is 0.352103 seconds.
Elapsed time is 0.348273 seconds.
Elapsed time is 0.353823 seconds.
Elapsed time is 0.352352 seconds.
Elapsed time is 0.366262 seconds.
Elapsed time is 0.372249 seconds.
```

```
Check IK pointwise with gaze target constraint  % master
Elapsed time is 0.464715 seconds.
Elapsed time is 0.373799 seconds.
Elapsed time is 0.379137 seconds.
Elapsed time is 0.344558 seconds.
Elapsed time is 0.358028 seconds.
Elapsed time is 0.350660 seconds.
Elapsed time is 0.361526 seconds.
Elapsed time is 0.346609 seconds.
Elapsed time is 0.359108 seconds.
Elapsed time is 0.358769 seconds.
Elapsed time is 0.360272 seconds.
Elapsed time is 0.354801 seconds.
Elapsed time is 0.350588 seconds.
Elapsed time is 0.355213 seconds.
Elapsed time is 0.358962 seconds.
Elapsed time is 0.363840 seconds.
Elapsed time is 0.367888 seconds.
Elapsed time is 0.360966 seconds.
Elapsed time is 0.362842 seconds.
Elapsed time is 0.365811 seconds.
```

Note that I found these tests to be frustratingly stochastic -- for instance whether or not matlab has focus seems to make a real difference.

In summary, I think this will resolve some spontaneous crash issues that people have been seeing on windows, and is in principle the right thing to do in the code.  I don't think it has a noticable impact on performance (might be slightly better on big problems, since the OS potentially doesn't have to swap the big hunk of memory back to the stack whenever running...).

Let me know what you think.

File: drake/solvers/NonlinearProgramSnoptmex.cpp [hunk 1]
```diff
@@ -60,6 +69,16 @@ void mexFunction(int nlhs, mxArray* plhs[], int nrhs, const mxArray* prhs[])
   {
     mexErrMsgIdAndTxt("Drake:NonlinearProgramSnoptmex:InvalidInputs","The usage is [x,F,info,xmul,Fmul] = NonlinearProgramSnoptmex(x,xlow,xupp,Flow,Fupp,userfun,ObjAdd,ObjRow,A,iAfun,jAvar,iGfun,jGvar,options)");
   }
+
+  if (lenrw==0) { // then initialize (sninit needs some default allocation)
+    lenrw = 500000;
+    rw.reset(new snopt::doublereal[lenrw]);
+    leniw = 500000;
+    iw.reset(new snopt::integer[leniw]);
+    lencw = 500;
```
Comment (psiorx) [inline drake/solvers/NonlinearProgramSnoptmex.cpp]: It's cleaner to roll the factor of 8 into this instead of multiplying it out many times later.

```C++```
lencw = 8 * 500;
```
or

```C++```
lencw = 500;
lencw_bytes = 8 * 500;
```
I count 29 instances of `8*lencw`

**Comment by psiorx** [comment]
:+1:

File: drake/solvers/NonlinearProgramSnoptmex.cpp (hunk 1 above)
Comment (RussTedrake) [inline drake/solvers/NonlinearProgramSnoptmex.cpp]: we need either the 8* or a /8 in places, because snopt takes it in different formats in different places.  i've taken a pass at cleaning it up before and gave up.  @hongkai-dai has too, i believe.  i agree it's gross, but will leave it as is for now.

File: drake/solvers/NonlinearProgramSnoptmex.cpp [hunk 2]
```diff
@@ -16,6 +17,14 @@ using namespace std;
...
 
+// NOTE: all snopt calls will use this shared memory... so this code is NOT THREAD SAFE
+static unique_ptr<snopt::doublereal []> rw;
+static unique_ptr<snopt::integer []> iw;
+static unique_ptr<char []> cw;
+static snopt::integer lenrw=0;
+static snopt::integer leniw=0;
+static snopt::integer lencw=0;
+
 static int snopt_userfun(snopt::integer *Status, snopt::integer *n, snopt::doublereal x[],
```
Comment (hongkai-dai) [inline drake/solvers/NonlinearProgramSnoptmex.cpp]: +1 for making them static. 

What is our preference in using pointer? Should we always prefer smart pointer to raw pointer?

Comment (RussTedrake) [inline drake/solvers/NonlinearProgramSnoptmex.cpp]: We should always avoid manual memory management.  That normally means smart pointers.
